# Contents:

1. Imports
2. Data Wrangling
3. Export

In [3]:
# Import libraries 

import pandas as pd
import numpy as np
import os

In [4]:
# create a path and import dataframe

path = r'C:\Users\walls\Documents\Coding\Data Analysis\CareerFoundry\Data Immersion A6\CA SAT Analysis'
df_sat = pd.read_csv(os.path.join(path, 'Data', 'Original Data', 'SAT Report 2015-2016.csv'), index_col = False)

In [5]:
df_sat.head()

index            cds rtype                                   sname  \
0      0              0     X                                     NaN   
1      1  1000000000000     C                                     NaN   
2      2  1100170000000     D                                     NaN   
3      3  1100170112607     S  Envision Academy for Arts & Technology   
4      4  1611190000000     D                                     NaN   

                                dname    cname  enroll12  NumTstTakr  \
0                                 NaN      NaN    492835      214262   
1                                 NaN  Alameda     16662        8611   
2  Alameda County Office of Education  Alameda       263          95   
3  Alameda County Office of Education  Alameda        88          92   
4                     Alameda Unified  Alameda       858         472   

  AvgScrRead AvgScrMath AvgScrWrit NumGE1500 PctGE1500  year  
0        484        494        477     89840     41.93  1516  
1        517        534        515      4696     54.53  1516  
2        395        378        388         6      6.32  1516  
3        391        376        386         5      5.43  1516  
4        527        543        514       301     63.77  1516

### Data Wrangling

In [6]:
df_sat.head()

index            cds rtype                                   sname  \
0      0              0     X                                     NaN   
1      1  1000000000000     C                                     NaN   
2      2  1100170000000     D                                     NaN   
3      3  1100170112607     S  Envision Academy for Arts & Technology   
4      4  1611190000000     D                                     NaN   

                                dname    cname  enroll12  NumTstTakr  \
0                                 NaN      NaN    492835      214262   
1                                 NaN  Alameda     16662        8611   
2  Alameda County Office of Education  Alameda       263          95   
3  Alameda County Office of Education  Alameda        88          92   
4                     Alameda Unified  Alameda       858         472   

  AvgScrRead AvgScrMath AvgScrWrit NumGE1500 PctGE1500  year  
0        484        494        477     89840     41.93  1516  
1        517        534        515      4696     54.53  1516  
2        395        378        388         6      6.32  1516  
3        391        376        386         5      5.43  1516  
4        527        543        514       301     63.77  1516

In [7]:
df_sat.shape

(2334, 14)

In [8]:
df_sat.tail()

index             cds rtype                                    sname  \
2329   2329  58727365830013     S                           Lindhurst High   
2330   2330  58727365830138     S  Marysville Charter Academy for the Arts   
2331   2331  58727365835202     S                          Marysville High   
2332   2332  58727690000000     D                                      NaN   
2333   2333  58727695838305     S                     Wheatland Union High   

                         dname cname  enroll12  NumTstTakr AvgScrRead  \
2329  Marysville Joint Unified  Yuba       249          74        430   
2330  Marysville Joint Unified  Yuba        38          21        532   
2331  Marysville Joint Unified  Yuba       202          56        493   
2332      Wheatland Union High  Yuba       168          60        500   
2333      Wheatland Union High  Yuba       168          60        500   

     AvgScrMath AvgScrWrit NumGE1500 PctGE1500  year  
2329        456        418        16     21.62  1516  
2330        500        513        13     61.90  1516  
2331        509        481        26     46.43  1516  
2332        496        481        27     45.00  1516  
2333        496        481        27     45.00  1516

In [9]:
df_sat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2334 entries, 0 to 2333
Data columns (total 14 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   index       2334 non-null   int64 
 1   cds         2334 non-null   int64 
 2   rtype       2334 non-null   object
 3   sname       1754 non-null   object
 4   dname       2276 non-null   object
 5   cname       2333 non-null   object
 6   enroll12    2334 non-null   int64 
 7   NumTstTakr  2334 non-null   int64 
 8   AvgScrRead  2132 non-null   object
 9   AvgScrMath  2132 non-null   object
 10  AvgScrWrit  2132 non-null   object
 11  NumGE1500   2132 non-null   object
 12  PctGE1500   2132 non-null   object
 13  year        2334 non-null   int64 
dtypes: int64(5), object(9)
memory usage: 255.4+ KB


In [10]:
df_sat.describe()

index           cds       enroll12     NumTstTakr    year
count  2334.000000  2.334000e+03    2334.000000    2334.000000  2334.0
mean   1166.500000  2.919475e+13     828.073265     366.907455  1516.0
std     673.912086  1.440837e+13   10706.572048    4708.943529     0.0
min       0.000000  0.000000e+00       0.000000       0.000000  1516.0
25%     583.250000  1.964725e+13      74.000000      13.000000  1516.0
50%    1166.500000  3.066597e+13     227.000000      89.000000  1516.0
75%    1749.750000  3.868478e+13     496.500000     227.750000  1516.0
max    2333.000000  5.872770e+13  492835.000000  214262.000000  1516.0

#### Observations: 
1. Column names unclear
2. cds should be a string
3. AvgScrRead, AvgScrMath, AvgScrWrit, NumGE1500, PctGE1500 should be int
4. year stats seem strange

In [19]:
# Renaming columns

df_sat.rename(columns = {'cds' : 'cds_code', 'rtype' : 'record_type', 'sname' : 'school', 'dname' : 'district', 'cname' : 'county', 'NumTstTakr' : 'test_takers', 'AvgScrRead' : 'avg_scr_read', 'AvgScrMath' : 'avg_scr_math', 'AvgScrWrit' : 'avg_scr_writ', 'NumGE1500' : 'scr>1500', 'PctGE1500' : 'percentage'}, inplace = True)

In [20]:
df_sat.head(1)

index  cds_code record_type school district county  enroll12  test_takers  \
0      0         0           X    NaN      NaN    NaN    492835       214262   

  avg_scr_read avg_scr_math avg_scr_writ scr>1500 percentage  year  
0          484          494          477    89840      41.93  1516

In [25]:
# Check for mixed types

for col in df_sat.columns.tolist():
  weird = (df_sat[[col]].map(type) != df_sat[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_sat[weird]) > 0:
    print (col)

school
district
county
avg_scr_read
avg_scr_math
avg_scr_writ
scr>1500
percentage


In [72]:
# Changing index school district county and year to string

df_sat[['index', 'school', 'district', 'county', 'year']] = df_sat[['index', 'school', 'district', 'county', 'year']].astype('str')

In [30]:
# Change cds_code to string

df_sat['cds_code'] = df_sat['cds_code'].astype('str')

In [49]:
# Previous error occured: ValueError: invalid literal for int() with base 10: '*'
# Changing * to NaN instead for now

df_sat[['avg_scr_read', 'avg_scr_math', 'avg_scr_writ', 'scr>1500', 'percentage']] = df_sat[['avg_scr_read', 'avg_scr_math', 'avg_scr_writ', 'scr>1500', 'percentage']].replace('*', np.nan)

In [54]:
# Converting new NaN values to numerics

columns_to_convert = ['avg_scr_read', 'avg_scr_math', 'avg_scr_writ', 'scr>1500', 'percentage']

# Apply to_numeric to each column

for col in columns_to_convert:
    df_sat[col] = pd.to_numeric(df_sat[col], errors='coerce')

In [57]:
df_sat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2334 entries, 0 to 2333
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   index         2334 non-null   int64  
 1   cds_code      2334 non-null   object 
 2   record_type   2334 non-null   object 
 3   school        2334 non-null   object 
 4   district      2334 non-null   object 
 5   county        2334 non-null   object 
 6   enroll12      2334 non-null   int64  
 7   test_takers   2334 non-null   int64  
 8   avg_scr_read  1731 non-null   float64
 9   avg_scr_math  1731 non-null   float64
 10  avg_scr_writ  1731 non-null   float64
 11  scr>1500      1731 non-null   float64
 12  percentage    1731 non-null   float64
 13  year          2334 non-null   int64  
dtypes: float64(5), int64(4), object(5)
memory usage: 255.4+ KB


In [58]:
# Checking for mixed types again

for col in df_sat.columns.tolist():
  weird = (df_sat[[col]].map(type) != df_sat[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_sat[weird]) > 0:
    print (col)

In [68]:
# Year as '1516' refers to 2015 to 2016 
# Will change to FY16 to refer to fiscal year starting in '15 and ending in '16

df_sat['year'] = df_sat['year'].replace('1516', 'FY16')

In [69]:
df_sat.head(5)

index       cds_code record_type                                  school  \
0      0              0           X                                     nan   
1      1  1000000000000           C                                     nan   
2      2  1100170000000           D                                     nan   
3      3  1100170112607           S  Envision Academy for Arts & Technology   
4      4  1611190000000           D                                     nan   

                             district   county  enroll12  test_takers  \
0                                 nan      nan    492835       214262   
1                                 nan  Alameda     16662         8611   
2  Alameda County Office of Education  Alameda       263           95   
3  Alameda County Office of Education  Alameda        88           92   
4                     Alameda Unified  Alameda       858          472   

   avg_scr_read  avg_scr_math  avg_scr_writ  scr>1500  percentage  year  
0         484.0         494.0         477.0   89840.0       41.93  FY16  
1         517.0         534.0         515.0    4696.0       54.53  FY16  
2         395.0         378.0         388.0       6.0        6.32  FY16  
3         391.0         376.0         386.0       5.0        5.43  FY16  
4         527.0         543.0         514.0     301.0       63.77  FY16

In [73]:
# Checking stats again

df_sat.describe()

enroll12    test_takers  avg_scr_read  avg_scr_math  avg_scr_writ  \
count    2334.000000    2334.000000   1731.000000   1731.000000   1731.000000   
mean      828.073265     366.907455    477.628538    482.608897    468.573079   
std     10706.572048    4708.943529     57.883539     61.802755     56.637737   
min         0.000000       0.000000    261.000000    264.000000    263.000000   
25%        74.000000      13.000000    436.000000    439.000000    429.500000   
50%       227.000000      89.000000    477.000000    482.000000    465.000000   
75%       496.500000     227.750000    518.500000    521.000000    505.000000   
max    492835.000000  214262.000000    657.000000    710.000000    677.000000   

           scr>1500   percentage  
count   1731.000000  1731.000000  
mean     207.079723    38.973131  
std     2260.998787    22.594265  
min        0.000000     0.000000  
25%       16.000000    20.435000  
50%       45.000000    37.040000  
75%      132.000000    55.850000  
max    89840.000000    98.530000

In [74]:
df_sat['year'].value_counts()

year
FY16    2334
Name: count, dtype: int64

##### Observations: 
1. Name changes good
2. Data type changes good
3. Year count looks good

In [75]:
# Checking for missing values

df_sat.isnull().sum()

index             0
cds_code          0
record_type       0
school            0
district          0
county            0
enroll12          0
test_takers       0
avg_scr_read    603
avg_scr_math    603
avg_scr_writ    603
scr>1500        603
percentage      603
year              0
dtype: int64

##### Observations:
1. 603 missing values for avg scr_read,scr_writ, scr_math, and percentage

In [81]:
df_sat[['avg_scr_read', 'avg_scr_math', 'avg_scr_writ', 'percentage']].isna().sum()

avg_scr_read    603
avg_scr_math    603
avg_scr_writ    603
percentage      603
dtype: int64

##### Observations: 
1. Missing values are the NaN values
2. Because they are equal in number for each column, going to keep them and look for any patterns

In [82]:
# Checking for duplicates 

df_dup = df_sat[df_sat.duplicated()]

In [83]:
df_dup

Empty DataFrame
Columns: [index, cds_code, record_type, school, district, county, enroll12, test_takers, avg_scr_read, avg_scr_math, avg_scr_writ, scr>1500, percentage, year]
Index: []

In [84]:
df_sat.shape

(2334, 14)

### 3. Export

In [87]:
# Export df_sat

df_sat.to_csv(os.path.join(path, 'Data','Prepared Data', 'SAT_Report_cleaned.csv'))